plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -360,6 +361,196 @@ struct block_time_tracker {
    bool paused = false;
 };
 
+/**
+ * Tracks active finalizers and votes via production_pause_vote_tracker to determine if implicit pause
+ * of production is needed.
+ */
+struct implicit_production_pause_vote_tracker {
+
+   void init(const controller& chain,
+             const std::set<chain::account_name>& producers,
+             fc::microseconds production_pause_vote_timeout)
+   {
+      _chain     = &chain;
+      _producers = producers; // copy, normally contains at most one
+
+      // Can be disabled either by configuration (if production_pause_vote_timeout.count() == 0)
+      // or because there can never be any possible production to pause (if producers.empty()).
+      // If it is not disabled for either of those two reasons, then change the
+      // initial vote_track_mode from its default value of `vote_track_mode::disabled`
+      // to `vote_track_mode::inactive` so that a future call to `update_active_finalizers`
+      // has an opportunity to set it correctly going forward.
+      if (!producers.empty() && production_pause_vote_timeout.count() != 0) {
+         _vt.set_vote_timeout(production_pause_vote_timeout);
+         _vote_track_mode.store(vote_track_mode::inactive, std::memory_order_relaxed);
+      }
+   }
+
+   enum class vote_track_mode {
+      disabled,
+      inactive,
+      only_other_votes,
+      only_producer_votes,
+      all_votes,
+   };
+
+   vote_track_mode get_vote_track_mode() const {
+      return _vote_track_mode.load(std::memory_order_relaxed);
+   }
+
+   static bool is_active(vote_track_mode vtm) {
+      return vtm != vote_track_mode::disabled && vtm != vote_track_mode::inactive;
+   }
+
+   // Called on resume()
+   void force_unpause() {
+      if (!is_active(get_vote_track_mode()))
+         return;
+
+      _vt.force_unpause(); // safe to always call, but no need if not active
+   }
+
+   auto check_pause_status(fc::time_point now) const {
+      const auto vtm = get_vote_track_mode();
+
+      production_pause_vote_tracker::pause_check check = production_pause_vote_tracker::pause_check::both;
+
+      switch (vtm) {
+      case vote_track_mode::disabled:
+      case vote_track_mode::inactive:
+         return production_pause_vote_tracker::pause_status{}; // should_pause() will return false
+      case vote_track_mode::only_other_votes:
+         check = production_pause_vote_tracker::pause_check::other;
+         break;
+      case vote_track_mode::only_producer_votes:
+         check = production_pause_vote_tracker::pause_check::producer;
+         break;
+      case vote_track_mode::all_votes:
+         check = production_pause_vote_tracker::pause_check::both;
+         break;
+      }
+
+      return _vt.check_pause_status(now, check);
+   }
+
+   // called from multiple threads
+   void on_vote(uint32_t connection_id, vote_result_t status, const vote_message_ptr& msg,
+                const finalizer_authority_ptr& active_auth, const finalizer_authority_ptr& pending_auth)
```

Comment:
Probably `active_auth` should be named as `active_finalizer` and`pending_auth` as `pending_finalizer`.

Reply:
These are `finalizer_authority`, maybe we could have named that type something else. We have a `finalizer` type, these are not those. I renamed to `active_finalizer_auth` and `pending_finalizer_auth`.